src/menu/Menu.ts

```diff
@@ -0,0 +1,435 @@
+ * @property animate			Determines whether animation should be handled internally.
+ * @property disabled			Determines whether the menu is disabled.
+ * @property expandOnClick		Determines whether a menu is displayed on click (default) or hover.
+ * @property hideDelay			The amount of time (in milliseconds) after mouseleave before hiding the menu.
+ * @property hidden				Determines whether the menu is hidden.
+ * @property hideOnActivate		Determines whether the menu should be hidden when an item is activated. Defaults to true.
+ * @property id					The widget ID. Defaults to a random string.
+ * @property label				A DNode to use as the trigger for a nested menu.
+ * @property nested				Indicates whether the menu is nested within another menu.
+ * @property onRequestHide		Called when the menu is displayed and the trigger is activated.
+ * @property onRequestShow		Called when the menu is hidden and the trigger is activated.
+ * @property role				The value to use for the menu's `role` property. Defaults to 'menu'.
+ */
+export interface MenuProperties extends ThemeableProperties {
+	active?: boolean;
+	animate?: boolean;
+	disabled?: boolean;
+	expandOnClick?: boolean;
+	hideDelay?: number;
+	hidden?: boolean;
+	hideOnActivate?: boolean;
+	id?: string;
+	label?: DNode;
+	nested?: boolean;
+	onRequestHide?: () => void;
+	onRequestShow?: () => void;
+	orientation?: Orientation;
+	role?: Role;
+}
+
+function getMenuHeight(menuElement: HTMLElement): number {
+	const maxHeight = parseInt(getComputedStyle(menuElement).getPropertyValue('max-height'), 10);
+	return Math.min(menuElement.scrollHeight, (isNaN(maxHeight) ? Infinity : maxHeight));
+}
+
+export const enum Operation {
+	decrease,
+	increase,
+	reset
+}
+
+const commonKeys = {
+	enter: 13,
+	escape: 27,
+	space: 32,
+	tab: 9
+};
+
+const horizontalKeys = assign({
+	ascend: 38, // up arrow
+	decrease: 37, // left arrow
+	descend: 40, // down arrow
+	increase: 39 // right arrow
+}, commonKeys);
+
+const verticalKeys = assign({
+	ascend: 37, // left arrow
+	decrease: 38, // up arrow
+	descend: 39, // right arrow
+	increase: 40 // down arrow
+}, commonKeys);
+
+export const MenuBase = StatefulMixin(ThemeableMixin(WidgetBase));
+
+@theme(css)
+export class Menu extends MenuBase<MenuProperties> {
+	protected wasOpen = false;
+	private _activeIndex = 0;
+	private _domNode: HTMLElement;
+	private _hideTimer: Handle;
+	private _id = uuid();
+	private _initialRender = true;
+	private _isLabelActive = false;
+
+	render(): DNode {
+		const {
+			id,
+			nested,
+			role = 'menu'
+		} = this.properties;
+
+		if (id) {
+			this._id = id;
+		}
+
+		const label = this.renderLabel();
+		const menu = v('div', {
+			classes: this.classes.apply(this, this.getMenuClasses()),
+			id: this._id,
+			key: 'menu',
+			onclick: this.onItemActivate,
+			onfocus: this.onMenuFocus,
+			onfocusout: this.onMenuFocusOut,
+			onkeydown: this.onMenuKeyDown,
+			role,
+			tabIndex: this.state.active || label ? -1 : 0
```

Comment:
Based on our private conversation, I can update `MenuItem` to store a reference to its DOM node so that focus can be moved out of `onElementUpdated` and changed in response to keyboard events. That way, this ~hack~ workaround can be removed.